test/createBaseStream.test.js

```diff
@@ -0,0 +1,232 @@
+
+      // when
+      stream.read()
+      await sleep(10)
+
+      // then
+      expect(touched).toBe(true)
+    })
+
+    test('calls the write function on write', async () => {
+      // given
+      let touched = false
+      const pipeline = { readable: true, writable: true }
+      const init = () => {}
+      const read = () => {}
+      const write = () => {
+        touched = true
+      }
+      const stream = createBaseStream(pipeline, { init, read, write })
+
+      // when
+      stream.write('')
+      await sleep(10)
+
+      // then
+      expect(touched).toBe(true)
+    })
+  })
+
+  describe('Readable', () => {
+    test('creates a Readable stream', async () => {
+      // given
+      const pipeline = { readable: true }
+      const init = () => {}
+      const read = () => {}
+
+      // when
+      const stream = createBaseStream(pipeline, { init, read })
+
+      // then
+      expect(isReadable(stream)).toBe(true)
+    })
+
+    test('calls the init function when read is called', async () => {
+      // given
+      let touched = false
+      const pipeline = { readable: true }
+      const init = () => {
+        touched = true
+      }
+      const read = () => {}
+      const stream = createBaseStream(pipeline, { init, read })
+
+      // when
+      stream.read()
+      await sleep(10)
+
+      // then
+      expect(touched).toBe(true)
+    })
+
+    test('calls the read function on read', async () => {
+      // given
+      let touched = false
+      const pipeline = { readable: true }
+      const init = () => {}
+      const read = () => {
+        touched = true
+      }
+      const stream = createBaseStream(pipeline, { init, read })
+
+      // when
+      stream.read()
+      await sleep(10)
+
+      // then
+      expect(touched).toBe(true)
+    })
+  })
+
+  describe('Writable', () => {
+    test('creates a Writable stream', async () => {
+      // given
+      const pipeline = { writable: true }
+      const init = () => {}
+      const write = () => {}
+
+      // when
+      const stream = createBaseStream(pipeline, { init, write })
+
+      // then
+      expect(isWritable(stream)).toBe(true)
+    })
+
+    test('calls the init function when write is called', async () => {
+      // given
+      let touched = false
+      const pipeline = { writable: true }
+      const init = () => {
+        touched = true
+      }
+      const write = () => {}
+      const stream = createBaseStream(pipeline, { init, write })
+
+      // when
+      stream.write('')
+      await sleep(10)
+
+      // then
+      expect(touched).toBe(true)
+    })
+
+    test('calls the write function on write', async () => {
+      // given
+      let touched = false
+      const pipeline = { writable: true }
+      const init = () => {}
+      const write = () => {
+        touched = true
+      }
+      const stream = createBaseStream(pipeline, { init, write })
+
+      // when
+      stream.write('')
+      await sleep(10)
```

Comment:
Why sleep and not stream events or something?

Reply:
Could be used. The idea was to not rely on anything besides the event loop.

Reply:
Well, I have been conditioned to consider any artificial delays in test code to be a bit of a smell. If there is inherent asynchronicity in the test code then I think we should reflect that in the tests too